### PR TITLE
Add background code for deploy function UI

### PR DESCRIPTION
**Pull request by Flanker32**
Add background code for deploy function UI

### CODE REVIEW

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/runner/functions/deploy/ui/FunctionDeploymentPanel.java [hunk 1]
```diff
@@ -84,15 +248,118 @@ protected JLabel getLblMavenProject() {
...
+    }
+
+    private AppServicePlan getAppServicePlan(FunctionApp functionApp) {
+        if (!appServicePlanMap.containsKey(functionApp.id())) {
+            appServicePlanMap.put(functionApp.id(), functionApp.manager().appServicePlans().getById(functionApp.appServicePlanId()));
+        }
+        return appServicePlanMap.get(functionApp.id());
+    }
+
+    private void selectFunction() {
```
Comment (andxu) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/runner/functions/deploy/ui/FunctionDeploymentPanel.java]: onFunctionSelected

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/runner/functions/deploy/ui/FunctionDeploymentPanel.java [hunk 2]
```diff
@@ -84,15 +248,118 @@ protected JLabel getLblMavenProject() {
...
+        configuration.setRuntime(runtimeConfiguration);
+
+        configuration.setPricingTier(appServicePlan.pricingTier().toSkuDescription().size());
+        configuration.setAppSettings(appSettingsTable.getAppSettings());
+        configuration.setModule((Module) cbFunctionModule.getSelectedItem());
+        configuration.setDeploymentStagingDirectory(txtStagingFolder.getText());
+    }
+
+    private AppServicePlan getAppServicePlan(FunctionApp functionApp) {
+        if (!appServicePlanMap.containsKey(functionApp.id())) {
```
Comment (andxu) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/runner/functions/deploy/ui/FunctionDeploymentPanel.java]: computeIfAbsent

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/runner/functions/deploy/ui/FunctionDeploymentPanel.java [hunk 3]
```diff
@@ -84,15 +248,118 @@ protected JLabel getLblMavenProject() {
...
+        presenter.loadFunctionApps(false);
     }
 
     @Override
     protected void apply(@NotNull FunctionDeployConfiguration configuration) {
+        if (selectedFunctionApp == null || selectedFunctionApp.getResource() == null) {
+            return;
+        }
+        final FunctionApp targetFunction = selectedFunctionApp.getResource();
+        configuration.setAppName(targetFunction.name());
```
Comment (andxu) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/runner/functions/deploy/ui/FunctionDeploymentPanel.java]: shall add a method accept FunctionApp  in configuration

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/runner/functions/deploy/ui/FunctionDeploymentPanel.java [hunk 4]
```diff
@@ -84,15 +248,118 @@ protected JLabel getLblMavenProject() {
...
+        @Override
+        public void customize(JList list, Object value, int index, boolean b, boolean b1) {
+            if (value == null) {
+                return;
+            } else if (value instanceof String) {
+                setText(getStringLabelText((String) value));
+            } else {
+                final ResourceEx<FunctionApp> function = (ResourceEx<FunctionApp>) value;
+                // For label in combobox textfield, just show webapp name
+                final String text = index >= 0 ? getFunctionAppLabelText(function.getResource()) : function.getResource().name();
```
Comment (andxu) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/runner/functions/deploy/ui/FunctionDeploymentPanel.java]: what's the case for negative index?

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/runner/functions/deploy/ui/FunctionDeploymentPanel.java [hunk 5]
```diff
@@ -84,15 +248,118 @@ protected JLabel getLblMavenProject() {
...
+
+        @Override
+        public void customize(JList list, Object value, int index, boolean b, boolean b1) {
+            if (value == null) {
+                return;
+            } else if (value instanceof String) {
+                setText(getStringLabelText((String) value));
+            } else {
+                final ResourceEx<FunctionApp> function = (ResourceEx<FunctionApp>) value;
+                // For label in combobox textfield, just show webapp name
```
Comment (andxu) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/runner/functions/deploy/ui/FunctionDeploymentPanel.java]: function

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/runner/functions/deploy/ui/FunctionDeploymentPanel.java (hunk 3 above)
Comment (andxu) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/runner/functions/deploy/ui/FunctionDeploymentPanel.java]: add a method in FunctionDeployConfiguration for FunctionApp

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/runner/functions/deploy/ui/FunctionDeploymentPanel.java [hunk 6]
```diff
@@ -84,15 +248,118 @@ protected JLabel getLblMavenProject() {
...
+        configuration.setFunctionId(targetFunction.id());
+
+        final IntelliJFunctionRuntimeConfiguration runtimeConfiguration = new IntelliJFunctionRuntimeConfiguration();
+        final AppServicePlan appServicePlan = getAppServicePlan(targetFunction);
+        runtimeConfiguration.setOs(appServicePlan.operatingSystem() == OperatingSystem.WINDOWS ? "windows" : "linux");
+        configuration.setRuntime(runtimeConfiguration);
+
+        configuration.setPricingTier(appServicePlan.pricingTier().toSkuDescription().size());
+        configuration.setAppSettings(appSettingsTable.getAppSettings());
+        configuration.setModule((Module) cbFunctionModule.getSelectedItem());
```
Comment (andxu) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/runner/functions/deploy/ui/FunctionDeploymentPanel.java]: shall we use RunConfigurationModule? see ModuleBasedConfiguration

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/runner/functions/deploy/ui/FunctionDeploymentPanel.java (hunk 1 above)
Comment (Flanker32) [reply PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/runner/functions/deploy/ui/FunctionDeploymentPanel.java]: Renamed

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/runner/functions/deploy/ui/FunctionDeploymentPanel.java (hunk 3 above)
Comment (Flanker32) [reply PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/runner/functions/deploy/ui/FunctionDeploymentPanel.java]: Done

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/runner/functions/deploy/ui/FunctionDeploymentPanel.java (hunk 4 above)
Comment (Flanker32) [reply PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/runner/functions/deploy/ui/FunctionDeploymentPanel.java]: This is for Label in editor, or the editor will be also shown as multi-line

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/runner/functions/deploy/ui/FunctionDeploymentPanel.java (hunk 5 above)
Comment (Flanker32) [reply PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/runner/functions/deploy/ui/FunctionDeploymentPanel.java]: Fixed

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/runner/functions/deploy/ui/FunctionDeploymentPanel.java (hunk 3 above)
Comment (Flanker32) [reply PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/runner/functions/deploy/ui/FunctionDeploymentPanel.java]: Done

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/runner/functions/deploy/ui/FunctionDeploymentPanel.java (hunk 6 above)
Comment (Flanker32) [reply PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/runner/functions/deploy/ui/FunctionDeploymentPanel.java]: `AzureRunConfigurationBase` is based on `RunConfigurationBase`, and other services like app services depends on project for now, we could investigate whether we could move to `ModuleBasedConfiguration` later

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/runner/functions/deploy/ui/FunctionDeploymentPanel.java [hunk 7]
```diff
@@ -1,45 +1,116 @@
...
+import javax.swing.JPanel;
+import javax.swing.event.PopupMenuEvent;
+import java.nio.file.Paths;
+import java.util.Arrays;
+import java.util.List;
 import java.util.Map;
 
+import static com.microsoft.intellij.runner.functions.AzureFunctionsConstants.EMPTY_TEXT;
+import static com.microsoft.intellij.runner.functions.AzureFunctionsConstants.LOADING_TEXT;
+
```
Comment (andxu) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/runner/functions/deploy/ui/FunctionDeploymentPanel.java]: two blank limes

**Pull request merged**